File: leetcode/AIcoding/3.DeepSeek-v3/go_solutions/1735-count-ways-to-make-array-with-product.go
const mod = 1e9 + 7
const maxK = 1e4 + 10
const maxN = 1e4 + 13

var (
    spf    [maxN]int
    comb   [maxN][15]int
    inited bool
)

func init() {
    if inited {
        return
    }
    inited = true
    sieve()
    precomputeComb()
}

func sieve() {
    for i := 2; i < maxN; i++ {
        if spf[i] == 0 {
            spf[i] = i
            if i*i < maxN {
                for j := i * i; j < maxN; j += i {
                    if spf[j] == 0 {
                        spf[j] = i
                    }
                }
            }
        }
    }
}

func precomputeComb() {
    comb[0][0] = 1
    for n := 1; n < maxN; n++ {
        comb[n][0] = 1
        for k := 1; k < 15; k++ {
            comb[n][k] = (comb[n-1][k-1] + comb[n-1][k]) % mod
        }
    }
}

func getPrimeFactors(x int) map[int]int {
    factors := make(map[int]int)
    for x > 1 {
        p := spf[x]
        for x%p == 0 {
            factors[p]++
            x /= p
        }
    }
    return factors
}

func waysToFillArray(queries [][]int) []int {
    res := make([]int, len(queries))
    for i, q := range queries {
        n, k := q[0], q[1]
        if k == 1 {
            res[i] = 1
            continue
        }
        factors := getPrimeFactors(k)
        ans := 1
        for _, cnt := range factors {
            ans = (ans * comb[n+cnt-1][cnt]) % mod
        }
        res[i] = ans
    }
    return res
}

File: leetcode/AIcoding/3.DeepSeek-v3/rust_solutions/1735-count-ways-to-make-array-with-product.rs
impl Solution {
    pub fn ways_to_fill_array(queries: Vec<Vec<i32>>) -> Vec<i32> {
        const MOD: i32 = 1_000_000_007;
        const MAX_N: usize = 10000 + 13;
        const MAX_K: usize = 10000 + 13;

        let mut min_prime = vec![0; MAX_K];
        for i in 2..MAX_K {
            if min_prime[i] == 0 {
                min_prime[i] = i as i32;
                for j in (i * i..MAX_K).step_by(i) {
                    if min_prime[j] == 0 {
                        min_prime[j] = i as i32;
                    }
                }
            }
        }

        let mut comb = vec![vec![0; 20]; MAX_N];
        for i in 0..MAX_N {
            comb[i][0] = 1;
            if i < 20 {
                comb[i][i] = 1;
            }
            for j in 1..std::cmp::min(i, 20) {
                comb[i][j] = (comb[i - 1][j - 1] + comb[i - 1][j]) % MOD;
            }
        }

        let mut res = Vec::with_capacity(queries.len());
        for query in queries {
            let (n, k) = (query[0] as usize, query[1] as usize);
            if k == 1 {
                res.push(1);
                continue;
            }
            let mut ans = 1i64;
            let mut x = k;
            while x > 1 {
                let p = min_prime[x] as usize;
                let mut cnt = 0;
                while x % p == 0 {
                    x /= p;
                    cnt += 1;
                }
                ans = (ans * comb[n + cnt - 1][cnt] as i64) % MOD as i64;
            }
            res.push(ans as i32);
        }
        res
    }
}